l10n_it_withholding_tax_payment/models/withholding_tax.py

```diff
@@ -56,7 +59,7 @@ def _compute_total(self):
         default=lambda self:
         self.env['res.company']._company_default_get('account.account'))
     name = fields.Char('Name')
-    date = fields.Char('Date')
+    date = fields.Date(string="Date")
```

Comment:
i dati vengono migrati o è necessario aggiungere uno script di migrazione?

Reply:
La rappresentazione fisica, all'interno del database, non dovrebbe cambiare; era un campo testuale contenente una data, rimane un campo testuale contenente una data.

https://github.com/OCA/l10n-italy/pull/1281/files#diff-306ae68589e8a31576b1a9cdfef4ccb2R159

---

L'unico dubbio che ho, riguarda questa riga: https://github.com/OCA/l10n-italy/pull/1281/files#diff-c5470233761e05b10ef57ae87db98002R61

Non vedo particolari logiche, anche nelle versioni precedenti, che impediscano di cambiare manualmente il valore del campo, all'interno della vista.
Inoltre, proprio perché il campo è stato di tipo `Char` non erano presenti particolari logiche di validazione sul formato della data.  

Nel caso in cui, nella versione precedente, un utente abbia modificato manualmente il campo inserendo un valore a sua discrezione, ahimè, c'è poco che si possa fare...

@alessandrocamilli, correggimi se sbaglio!

---

Ci si potrebbe mettere a parsare tutti i campi data usando *N* formati "comuni" di date cercando di matchare il dato, per poi riscriverlo nel formato standard... 
L'alternativa, potrebbe essere una migrazione che "pulisca" i campi laddove non sia possibile parsare la data usando il formato standard, per evitare errori lato ORM?
  
Anche se, non mi piace tantissimo rimuovere informazioni da un DB... 🤔

---

Soluzioni?
Potrebbe essere un caso da gestire?
Lo ignoriamo?

Reply:
> La rappresentazione fisica, all'interno del database, non dovrebbe cambiare; era un campo testuale contenente una data, rimane un campo testuale contenente una data.
> 
> https://github.com/OCA/l10n-italy/pull/1281/files#diff-306ae68589e8a31576b1a9cdfef4ccb2R159

Ok grazie per la spiegazione. Mi chiedo però cosa succeda se la stringa non è nel formato che si aspetta, che se non erro è (YYYY-MM-DD).

Reply:
>  Mi chiedo però cosa succeda se la stringa non è nel formato che si aspetta [...]

Sissì, ti ho espresso lo stesso dubbio, infatti...
Come facevo presente, non saprei come gestirla nella maniera ottimale...

-  Prendo in considerazioni più formati (`%Y-%m-%d` ,`%y/%m/%d`, `%d/%m/%Y`, `%b/%d/%Y`, ecc...) e tento di parsarle, uno dopo l'altro?
-  Se falliscono tutti i parsing, elimino il vecchio valore?
-  È un problema che si può ignorare?

---

@eLBati?  
Idee?

Reply:
Effettivamente c'è la possibilità che nel campo ci sia qualcosa che non sia una data e che quindi possa spaccare il modulo.

Volendo mitigare questo problema mi limiterei però a buttare via tutto ciò che non è una data `YYYY-MM-DD`: se l'utente è andato a modificare manualmente il campo inserendo cose strane, se l'è andata a cercare

Reply:
Aggiunta la migration! 😉

Reply:
Ottimo!
Sei riuscito anche a testarla?

Reply:
Yep! 😉

Anche se...  
Non ho fisicamente messo in piedi un'istanza `v11`, su cui eseguire un'intera migrazione alla `v12` per testare questa singola migration...

Ho, semplicemente, alterato via SQL la singola tabella `withholding_tax_move_payment` settando la colonna `date` ad un `CHARACTER VARYING` e valorizzando alcune righe con stringhe non parsabili come date.  
Dopodiché ho lanciato l'update e la conseguente migration è stata correttamente eseguita.

```log
2019-12-05 14:52:44,551 1 INFO odoo12community odoo.modules.migration: module l10n_it_withholding_tax_payment: Running migration [>12.0.1.0.0] pre-migrate
2019-12-05 14:52:44,567 1 ERROR odoo12community pre-migrate: The row in table 'withholding_tax_move_payment' with ID #2 and 'date' field valued as '2019-12-05 asd' will be set as NULL...
2019-12-05 14:52:44,570 1 ERROR odoo12community pre-migrate: The row in table 'withholding_tax_move_payment' with ID #3 and 'date' field valued as '' will be set as NULL...
2019-12-05 14:52:44,571 1 ERROR odoo12community pre-migrate: The row in table 'withholding_tax_move_payment' with ID #4 and 'date' field valued as 'None' will be set as NULL...
2019-12-05 14:52:44,573 1 ERROR odoo12community pre-migrate: The row in table 'withholding_tax_move_payment' with ID #5 and 'date' field valued as 'asdasdasd' will be set as NULL...
2019-12-05 14:52:44,639 1 INFO odoo12community pre-migrate: Migration executed successfully. Exiting...
```

---

Mi è sembrato sufficiente, come test... 🤔

Reply:
Anche secondo me è sufficiente

Reply:
Attendiamo, quindi, l'approvazione da parte di @tafaRU per poter continuare con il merge...